### ISSUE TITLE
examples about how to use dali multi gpu in pytorch 

### ISSUE THREAD

**Issue by songtaoshi** [NVIDIA/DALI#2148]
Hi I have seen the sharding in the tutorial, I know there are different pipelines. But I am wondering how to write a full pytorch script? Is is like we need to use local rank to determine the allocation of pipeline to the GPU

**Comment by JanuszL** [comment]
Hi,
Yes, you need a local rand to assign the proper GPU, and global rank to assign the proper shard. Please take a look at this [RN50 PyTorch example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/use_cases/pytorch/resnet50/pytorch-resnet50.html).

**Comment by songtaoshi** [comment]
Hi, the example using ops.FileReader, it is very convenient to do the sharding, but I am wondering how to implement my own inputer, but I think the logic is the same, just sharding the data to the ranks.  and the gradient are connected by apex.distributed, Do I understand right?

**Comment by JanuszL** [comment]
Hi,
I recommend using ExternalSource operator like in [this example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/external_input.html).
>  the logic is the same, just sharding the data to the ranks

Yes, that is what should be done.

**Comment by songtaoshi** [comment]
Hi, It seems that the dataloader is unbalanced. I have nearly same sharding data to each GPU, But the GPU memory usage is not the same.

**Comment by songtaoshi** [comment]
![image](https://user-images.githubusercontent.com/20240391/89489579-688c7080-d7dd-11ea-970a-1acd9f2ac7bf.png)
![image](https://user-images.githubusercontent.com/20240391/89489600-77732300-d7dd-11ea-8626-d81bcfb125f2.png)
![image](https://user-images.githubusercontent.com/20240391/89489644-8bb72000-d7dd-11ea-9699-7e4e99c0ac9c.png)
I think I have use the same sharding, why do the gpu memory is unbalanced

**Comment by songtaoshi** [comment]
I use pytorch distributed training framework. 
![image](https://user-images.githubusercontent.com/20240391/89489779-d6d13300-d7dd-11ea-9ab0-9e0ae2621a00.png)

**Comment by JanuszL** [comment]
Are you sure that you run the model in each GPU? DALI should not consume whole GPU memory (16GB in your case) on its own. Also in the screen you provide I see a magnitude of the difference 4700 vs 40000 of file lengths.
Can you run your code on one GPU at the time and go by each GPU one by one (like run everything on the GPU no.2 only)?

**Comment by songtaoshi** [comment]
Hi the 4700 vs 40000 is dev dataloader vs training dataloader.  

each rank allocated 4700 dev and 40000 train 

local rank = 0 
![image](https://user-images.githubusercontent.com/20240391/89706456-67537300-d998-11ea-9973-28ebb7dd1162.png)

local rank = 1
![image](https://user-images.githubusercontent.com/20240391/89706487-c74a1980-d998-11ea-8f00-6f484ad4e659.png)

local rank = 2
![image](https://user-images.githubusercontent.com/20240391/89706509-f8c2e500-d998-11ea-8b83-e96fa0016d76.png)

**Comment by songtaoshi** [comment]
I think I want to express that I have almost the same sharding on each GPU, but the usage is unbalanced. 

I have tried with sharding data version, running with different local rank with only one GPU in the last comment.  

This comment, I have tried to change data num allocated to local rank 0, to a small scale. Besides,  I comment the sharding part, also running with different local rank with only one GPU. 

![image](https://user-images.githubusercontent.com/20240391/89706707-c1edce80-d99a-11ea-8c64-5959a5fe791d.png)

![image](https://user-images.githubusercontent.com/20240391/89706882-44c35900-d99c-11ea-87df-6b7fb5d05058.png)

The usage is the same. So I have some doubts that if I shard the data, there is an unbalance with dali dataloader

**Comment by songtaoshi** [comment]
My dataloader code.  
```python  
from __future__ import division
from nvidia.dali.pipeline import Pipeline
from nvidia.dali.plugin.pytorch import DALIGenericIterator
import nvidia.dali.ops as ops
import nvidia.dali.types as types
import kaldiio
from random import shuffle
import json
import numpy as np

'''
still things to be optimized
'''

FEATS_PAD = 0
LABELS_PAD = 2

class kaldi_input_iterator(object):
    def __init__(self, batch_size=1, device_id=0, num_gpus=1,data_json_path='',debug=False, start=0,end=200):

        with open(data_json_path, 'rb') as f:
            data = json.load(f)['utts']
        # 但是这种sort方式可以进行优化
        # sort it by input lengths (long to short)
        sorted_data = sorted(data.items(), key=lambda data: int(
            data[1]['input'][0]['shape'][0]), reverse=True)
        if debug == True:
            sorted_data = sorted_data[start:end]

        # whole data set size
        self.data_set_len = len(sorted_data) 

        print("please check the num of data to see if it is in debug mode",
              self.data_set_len)
        # based on the device_id and total number of GPUs - world size
        # get proper shard
        # self.files = sorted_data[self.data_set_len * device_id // num_gpus:
        #                         self.data_set_len * (device_id + 1) // num_gpus]

        self.files = sorted_data[:2000]

        # self.files = sorted_data
        self.batch_size = batch_size
        self.n = len(self.files)
        print('this is the length of the file',self.n)

    def __iter__(self):
        self.i = 0
        shuffle(self.files)
        return self

    def __next__(self):
        feats = []
        labels = []

        if self.i >= self.n:
            raise StopIteration

        for _ in range(self.batch_size):
            feat_path = self.files[self.i][1]['input'][0]['feat']
            label = self.files[self.i][1]['output'][0]['tokenid'].split()
            feat = kaldiio.load_mat(feat_path)
            # to be otimized
            # 以后弄出来的numpy得是C contiguous
            feats.append(np.ascontiguousarray(feat))
            # feats.append(np.ascontiguousarray(np.expand_dims(feat, -1)))
            # feats.append(feat)
            labels.append(np.array(label, dtype=np.int64))
            self.i = (self.i + 1) % self.n
        return (feats, labels)

    @property
    def size(self,):
        return self.data_set_len

    next = __next__


class train_pipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, external_data, axes=[0,1], nregions=8,target_max_len=60):
        super(train_pipeline, self).__init__(batch_size,
                                                     num_threads,
                                                     device_id,
                                                     seed=12)
        self.input = ops.ExternalSource()
        self.input_label = ops.ExternalSource()
        self.external_data = external_data
        self.iterator = iter(self.external_data)

        #padding implementation
        #since we are using pytorch iterator, so we need make the output in the uniform
        self.feats_pad = ops.Pad(fill_value=FEATS_PAD)
        self.labels_pad = ops.Pad(fill_value=LABELS_PAD, shape=target_max_len)

        # SpecAug
        ndims = len(axes)
        args_shape = (ndims*nregions)
        self.random_anchor = ops.Uniform(range=(0., 1.), shape=args_shape)
        self.random_shape = ops.Uniform(range=(0., 0.2), shape=args_shape)
        self.erase = ops.Erase(device="cpu",
                               axes=axes,
                               fill_value=0,
                               normalized_anchor=True,
                               normalized_shape=True)


    def define_graph(self):
        self.feats = self.input()
        self.labels = self.input_label()
        feats_aug = self.erase(self.feats, anchor=self.random_anchor(), shape=self.random_shape())
        feats_pd = self.feats_pad(feats_aug)
        labels_pd = self.labels_pad(self.labels)
        return (feats_pd, labels_pd)

    def iter_setup(self):
        try:
            (feats, labels) = self.iterator.next()
            self.feed_input(self.feats, feats)
            self.feed_input(self.labels, labels)
        except StopIteration:
            self.iterator = iter(self.external_data)
            raise StopIteration


class dev_pipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, external_data, target_max_len=60):
        super(dev_pipeline, self).__init__(batch_size,
                                                     num_threads,
                                                     device_id,
                                                     seed=12)
        self.input = ops.ExternalSource()
        self.input_label = ops.ExternalSource()
        self.external_data = external_data
        self.iterator = iter(self.external_data)

        #padding implementation
        #since we are using pytorch iterator, so we need make the output in the uniform
        self.feats_pad = ops.Pad(fill_value=FEATS_PAD)
        self.labels_pad = ops.Pad(fill_value=LABELS_PAD, shape=target_max_len)

    def define_graph(self):
        self.feats = self.input()
        self.labels = self.input_label()
        feats_pd = self.feats_pad(self.feats)
        labels_pd = self.labels_pad(self.labels)
        return (feats_pd, labels_pd)

    def iter_setup(self):
        try:
            (feats, labels) = self.iterator.next()
            self.feed_input(self.feats, feats)
            self.feed_input(self.labels, labels)
        except StopIteration:
            self.iterator = iter(self.external_data)
            raise StopIteration



def make_train_pipeline(batch_size=32, device_id=0, num_gpus=1, data_json_path='', num_threads=2):
    eii = kaldi_input_iterator(batch_size=batch_size, device_id=device_id, num_gpus=num_gpus,
                               data_json_path=data_json_path,start=0,end=200)

    pipe = train_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id,
                                external_data=eii)

    #这里需要确保同一个batch的size是统一的
    # dynamic shape就是可以让我不同的batch之间的shape不一样
    dataloader = DALIGenericIterator(
                                pipe, output_map=['feats','labels'],size=eii.size, 
                                auto_reset=True, fill_last_batch=True, 
                                dynamic_shape=True, last_batch_padded=False)
    return dataloader


def make_dev_pipeline(batch_size=32, device_id=0, num_gpus=1, data_json_path='', num_threads=2):
    eii = kaldi_input_iterator(batch_size=batch_size, device_id=device_id, num_gpus=num_gpus,
                               data_json_path=data_json_path,start=200,end=300)

    pipe = dev_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id,
                          external_data=eii)

    #这里需要确保同一个batch的size是统一的
    # dynamic shape就是可以让我不同的batch之间的shape不一样
    dataloader = DALIGenericIterator(
        pipe, output_map=['feats', 'labels'], size=eii.size,
        auto_reset=True, fill_last_batch=False,
        dynamic_shape=True, last_batch_padded=False)
    return dataloader

```

**Comment by songtaoshi** [comment]
@JanuszL really hope to your answer!

**Comment by JanuszL** [comment]
Hi,
I would check the size of the samples you have in each pipeline. Maybe for some reason, the first one gets the biggest one (as I see you are sorting them by size, right?)?
You may also want to check [this function](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/pipeline.html#nvidia.dali.pipeline.Pipeline.executor_statistics) and see if each operator's output buffer consumes the same amount of memory.
You can also reduce the batch size to 1 and compare this one sample from each pipeline side by side.

**Comment by songtaoshi** [comment]
Thanks for your reply. I notice that this is due to different shape of each sharding data, they padding to different size. Thanks a lot!
